### PR TITLE
maxBytesPerRead channel configuration

### PR DESCRIPTION
**Pull request by Scottmitch**
Motiviation:
The current read loops don't fascilitate reading a maximum amount of bytes. This capability is useful to have more fine grain control over how much data is injested.

Modifications:
- Add a setMaxBytesPerRead(int) and getMaxBytesPerRead() to ChannelConfig
- Add a setMaxBytesPerIndividualRead(int) and getMaxBytesPerIndividualRead to ChannelConfig
- Add methods to RecvByteBufAllocator so that a pluggable scheme can be used to control the behavior of the read loop.
- Modify read loop for all transport types to respect the new RecvByteBufAllocator API

Result:
The ability to control how many bytes are read for each read operation/loop, and a more extensible read loop.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@trustin - Would you mind reviewing this one?

File: transport/src/main/java/io/netty/channel/DefaultChannelConfig.java [hunk 1]
```diff
@@ -227,6 +258,85 @@ public ChannelConfig setMaxMessagesPerRead(int maxMessagesPerRead) {
...
+        return this;
+    }
+
+    @Override
+    public synchronized Entry<Integer, Integer> getMaxBytesPerReadPair() {
+        return new AbstractMap.SimpleEntry<Integer, Integer>(maxBytesPerRead, maxBytesPerIndividualRead);
+    }
+
+    @Override
+    public ChannelConfig setMaxBytesPerReadPair(int maxBytesPerRead, int maxBytesPerIndividualRead) {
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelConfig.java]: I think something similar can be done to resolve https://github.com/netty/netty/issues/3806.

File: transport/src/main/java/io/netty/channel/DefaultChannelConfig.java [hunk 2]
```diff
@@ -167,6 +190,14 @@ public boolean setOptions(Map<ChannelOption<?>, ?> options) {
...
         } else if (option == MAX_MESSAGES_PER_READ) {
             setMaxMessagesPerRead((Integer) value);
+        } else if (option == MAX_BYTES_PER_READ) {
+            setMaxBytesPerRead((Integer) value);
+        } else if (option == MAX_BYTES_PER_INDIVIDUAL_READ) {
+            setMaxBytesPerIndividualRead((Integer) value);
+        } else if (option == MAX_BYTES_PER_READ_PAIR) {
+            @SuppressWarnings("unchecked")
+            Entry<Integer, Integer> pair = (Entry<Integer, Integer>) value;
+            setMaxBytesPerReadPair(pair.getKey(), pair.getValue());
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelConfig.java]: I think something similar can be done to resolve #3806.

File: codec-dns/src/test/java/io/netty/handler/codec/dns/package-info.java [hunk 3]
```diff
@@ -1,21 +0,0 @@
...
- * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
- * License for the specific language governing permissions and limitations
- * under the License.
- */
-
-/**
- * A DNS client that queries a server and checks if query information and
- * responses are valid to ensure codec is correct.
- */
-package io.netty.handler.codec.dns;
```
Comment (trustin) [inline codec-dns/src/test/java/io/netty/handler/codec/dns/package-info.java]: ???

**Comment by trustin** [comment]
@Scottmitch Sorry for a delay. Will review some time this week.

File: codec-dns/src/test/java/io/netty/handler/codec/dns/package-info.java (hunk 3 above)
Comment (Scottmitch) [inline codec-dns/src/test/java/io/netty/handler/codec/dns/package-info.java]: Maybe doesn't belong in this PR, but this is a duplicate of the `package-info.java` in the `main` directory tree.

**Comment by Scottmitch** [comment]
@netkins build

**Comment by Scottmitch** [comment]
@trustin - Bump.

File: codec-dns/src/test/java/io/netty/handler/codec/dns/package-info.java (hunk 3 above)
Comment (trustin) [inline codec-dns/src/test/java/io/netty/handler/codec/dns/package-info.java]: Ah, nevermind, then.

**Comment by trustin** [comment]
Do the new configuration properties work only when the `RecvByteBufAllocator` is aware of them? If so, why don't we move them to the `RecvByteBufAllocator` implementation?

**Comment by Scottmitch** [comment]
Yes. Currently `MaxBytesRecvByteBufAllocator` is the only consumer of the new properties.  The `RecvByteBufAllocator` implementations have also encapsulated the responsibility of the existing `getMaxMessagesPerRead()`.  I think if we move the new parameters it may make sense to also move `getMaxMessagesPerRead()` into the respective implementation that uses it (this change is only for 4.1 and 5.0).  WDYT?

**Comment by Scottmitch** [comment]
@trustin - Another thing to consider is moving these configuration options from `ChannelConfig` to an implementation of `RecvByteBufAllocator` may make it a bit more cumbersome to change on the fly. For example this may require conditionals based upon the run time type of the `RecvByteBufAllocator`. For example:

```java
RecvByteBufAllocator recvAlloc = channel.config().getRecvByteBufAllocator();
if (recvAlloc instanceof MaxBytesRecvByteBufAllocator) {
  MaxBytesRecvByteBufAllocator maxRecvAlloc = (MaxBytesRecvByteBufAllocator) recvAlloc;
  maxRecvAlloc.setMaxMessagesPerRead(...);
  maxRecvAlloc.setMaxBytesPerIndividualRead(...);
} else if (recvAlloc instanceof ...) {
}
```

**Comment by trustin** [comment]
> I think if we move the new parameters it may make sense to also move `getMaxMessagesPerRead()` into the respective implementation that uses it (this change is only for 4.1 and 5.0). WDYT?

Sounds good to me!

> Another thing to consider is moving these configuration options from `ChannelConfig` to an implementation of `RecvByteBufAllocator` may make it a bit more cumbersome to change on the fly.

Yeah, how about using the type parameter for the return type for loose typing to lessen the burden?

```java
interface ChannelConfig {
    <T extends RecvByteBufAllocator> T getRecvByteBufAllocator();
}

MaxBytesRecvByteBufAllocator alloc = ch.config().getRecvByteBufAllocator();
...
```

This won't work when a user is not sure about the type of the current `RecvByteBufAllocator`, though...

**Comment by Scottmitch** [comment]
@trustin - Thanks for the feedback. I will move forward with these updates and ping you when ready.

File: transport-sctp/src/main/java/io/netty/channel/sctp/oio/OioSctpChannel.java [hunk 4]
```diff
@@ -184,8 +185,8 @@ protected int doReadMessages(List<Object> msgs) throws Exception {
 
         Set<SelectionKey> reableKeys = readSelector.selectedKeys();
         try {
-            for (SelectionKey ignored : reableKeys) {
-                RecvByteBufAllocator.Handle allocHandle = unsafe().recvBufAllocHandle();
+            for (@SuppressWarnings("unused") SelectionKey ignored : reableKeys) {
```
Comment (Scottmitch) [inline transport-sctp/src/main/java/io/netty/channel/sctp/oio/OioSctpChannel.java]: Not sure why this loop is necessary...see https://github.com/netty/netty/issues/3884

**Comment by Scottmitch** [comment]
@trustin @normanmaurer - PTAL.

File: transport-sctp/src/main/java/io/netty/channel/sctp/oio/OioSctpChannel.java (hunk 4 above)
Comment (trustin) [inline transport-sctp/src/main/java/io/netty/channel/sctp/oio/OioSctpChannel.java]: I'm not sure, either. Could you explain, @jestan ?

**Comment by trustin** [comment]
@Scottmitch LGTM. @normanmaurer, any comments?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java [hunk 5]
```diff
@@ -58,11 +58,6 @@
 
         return true;
     }
-    @Override
-    public EpollDomainSocketChannelConfig setMaxMessagesPerRead(int maxMessagesPerRead) {
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java]: I'm right that we only want this for 4.1 and master ?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorMessageHandle.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+final class EpollRecvByteAllocatorMessageHandle extends EpollRecvByteAllocatorHandle {
+    public EpollRecvByteAllocatorMessageHandle(RecvByteBufAllocator.Handle handle, boolean isEdgeTriggered) {
+        super(handle, isEdgeTriggered);
+    }
+
+    @Override
+    public boolean continueReading() {
+        // if edgeTriggered is used we need to read all bytes/messages as we are not notified again otherwise.
+        // For packet oriented descriptors must read until we get a EAGAIN
+        // (see Q9 in http://man7.org/linux/man-pages/man7/epoll.7.html).
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorMessageHandle.java]: Just a nit but we could use javadoc style and use `<a href="..."> ...`

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorStreamingHandle.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+final class EpollRecvByteAllocatorStreamingHandle extends EpollRecvByteAllocatorHandle {
+    public EpollRecvByteAllocatorStreamingHandle(RecvByteBufAllocator.Handle handle, boolean isEdgeTriggered) {
+        super(handle, isEdgeTriggered);
+    }
+
+    @Override
+    public boolean continueReading() {
+        // if edgeTriggered is used we need to read all bytes/messages as we are not notified again otherwise.
+        // For stream oriented descriptors we can assume we are done reading if the last read attempt didn't produce
+        // a full buffer (see Q9 in http://man7.org/linux/man-pages/man7/epoll.7.html).
```
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorStreamingHandle.java]: same comment as above

File: transport/src/main/java/io/netty/channel/AdaptiveRecvByteBufAllocator.java [hunk 8]
```diff
@@ -84,14 +81,14 @@ private static int getSizeTableIndex(final int size) {
...
-    private static final class HandleImpl implements Handle {
+    private final class HandleImpl extends MaxMessageHandle {
         private final int minIndex;
         private final int maxIndex;
         private int index;
         private int nextReceiveBufferSize;
         private boolean decreaseNow;
 
-        HandleImpl(int minIndex, int maxIndex, int initial) {
+        public HandleImpl(int minIndex, int maxIndex, int initial) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/AdaptiveRecvByteBufAllocator.java]: why public ?

File: transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+package io.netty.channel;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+
+public abstract class DefaultMaxMessagesRecvByteBufAllocator implements MaxMessagesRecvByteBufAllocator {
+    private volatile int maxMessagesPerRead;
+
+    public DefaultMaxMessagesRecvByteBufAllocator() {
+        // TODO(scott): this was previously based upon channel type, do we still need to consider channel type?
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java]: @trustin WDYT about this ?

File: transport/src/main/java/io/netty/channel/FixedRecvByteBufAllocator.java [hunk 10]
```diff
@@ -15,36 +15,23 @@
...
-public class FixedRecvByteBufAllocator implements RecvByteBufAllocator {
-
-    private static final class HandleImpl implements Handle {
+public class FixedRecvByteBufAllocator extends DefaultMaxMessagesRecvByteBufAllocator {
 
+    private final class HandleImpl extends MaxMessageHandle {
         private final int bufferSize;
 
-        HandleImpl(int bufferSize) {
+        public HandleImpl(int bufferSize) {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/FixedRecvByteBufAllocator.java]: why this change ?

File: transport/src/main/java/io/netty/channel/MaxBytesRecvByteBufAllocator.java [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+import java.util.Map.Entry;
+
+public interface MaxBytesRecvByteBufAllocator extends RecvByteBufAllocator {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/MaxBytesRecvByteBufAllocator.java]: Class hierarchy javadocs please

File: transport/src/main/java/io/netty/channel/MaxMessagesRecvByteBufAllocator.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+public interface MaxMessagesRecvByteBufAllocator extends RecvByteBufAllocator {
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/MaxMessagesRecvByteBufAllocator.java]: class hierarchy javadocs please

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java [hunk 13]
```diff
@@ -283,6 +283,8 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
             throw new NotYetConnectedException();
         case CLOSED:
             throw new ClosedChannelException();
+        case CONNECTED:
+            // We are good to write
```
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: add break; ?

**Comment by normanmaurer** [comment]
@Scottmitch a few more comments and questions. Good work!

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java (hunk 5 above)
Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java]: Yes. https://github.com/netty/netty/pull/3885#issuecomment-117373563.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorMessageHandle.java (hunk 6 above)
Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorMessageHandle.java]: Done

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorStreamingHandle.java (hunk 7 above)
Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorStreamingHandle.java]: Done

File: transport/src/main/java/io/netty/channel/AdaptiveRecvByteBufAllocator.java (hunk 8 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/AdaptiveRecvByteBufAllocator.java]: The class is `private` however there is no reason why the visibility of this method should be restricted.

File: transport/src/main/java/io/netty/channel/FixedRecvByteBufAllocator.java (hunk 10 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/FixedRecvByteBufAllocator.java]: The class is `private` however there is no reason why the visibility of this method should be restricted.

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java (hunk 5 above)
Comment (normanmaurer) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java]: I'm still not 100 % convinced that we can do this change in 4.1 in terms of API breakage. @trustin WDYT ?

File: transport/src/main/java/io/netty/channel/MaxBytesRecvByteBufAllocator.java (hunk 11 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/MaxBytesRecvByteBufAllocator.java]: Done

File: transport/src/main/java/io/netty/channel/MaxMessagesRecvByteBufAllocator.java (hunk 12 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/MaxMessagesRecvByteBufAllocator.java]: Done

File: transport/src/main/java/io/netty/channel/local/LocalChannel.java (hunk 13 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/local/LocalChannel.java]: Sure.

File: transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java (hunk 9 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java]: @trustin - Yes it would be good to get your input on this one.

**Comment by Scottmitch** [comment]
@trustin - I would like to merge this but there are just a few followup discussion items above.  Would you mind weighing in?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java (hunk 5 above)
Comment (trustin) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java]: @normanmaurer Yeah, we can't just remove it. How about deprecating it and delegate to `Max*Allocator` if the current recv allocator is of compatible type, or throw an exception if it's incompatible?

File: transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java (hunk 9 above)
Comment (trustin) [inline transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java]: That's a great question. With this PR, the accept throughput would be decreased due to excessive emission of `channelReadComplete()` events for `ServerSocketChannel`.  How about adding some protected abstract method so that the config implementations create the right recv buf allocator?

File: transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java (hunk 5 above)
Comment (Scottmitch) [inline transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollDomainSocketChannelConfig.java]: @trustin @normanmaurer - Done.

File: transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java (hunk 9 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java]: @trustin - I don't think adding a protected abstract method to the config fits in with the intended use case.  The user should have the flexibility to swap in what ever `RecvByteBufAllocator` they want independent of the channel or configuration.  I will add a field to `ChannelMetadata` to enforce a minimum value instead, and let plz let me know what you think.

**Comment by Scottmitch** [comment]
@trustin @normanmaurer - PTAL.

File: transport/src/main/java/io/netty/channel/DefaultChannelConfig.java [hunk 14]
```diff
@@ -255,16 +275,44 @@ public ChannelConfig setAllocator(ByteBufAllocator allocator) {
...
+     */
+    public ChannelConfig setRecvByteBufAllocator(RecvByteBufAllocator allocator, ChannelMetadata metadata) {
         if (allocator == null) {
             throw new NullPointerException("allocator");
         }
+        if (allocator instanceof MaxMessagesRecvByteBufAllocator) {
+            if (metadata == null) {
+                throw new NullPointerException("metadata");
+            }
+            MaxMessagesRecvByteBufAllocator maxMsgAllocator = (MaxMessagesRecvByteBufAllocator) allocator;
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/DefaultChannelConfig.java]: @trustin - This is how I am preserving the 16 max messages for `ServerSockets`/`NioByteStream` and 1 max message otherwise.  WDYT?

Comment (trustin) [inline transport/src/main/java/io/netty/channel/DefaultChannelConfig.java]: LGTM. Thank you for your patience.

**Comment by Scottmitch** [comment]
@trustin - I guess now that this doesn't break any APIs we could also consider putting this in 4.0. It would be nice to be consistent between 4.0 and 4.1...WDYT?

**Comment by Scottmitch** [comment]
Ran some quick perf tests on the before and after for this PR just to make sure there was no regression, and things look good (using NIO):

before
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    22.07ms   33.58ms   1.11s    85.72%
    Req/Sec   132.23k    33.88k  241.31k    66.05%
  251377804 requests in 2.00m, 33.24GB read
Requests/sec: 2093223.99
Transfer/sec:    283.47MB
```

this PR
```
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    21.36ms   34.33ms   1.28s    85.67%
    Req/Sec   136.39k    29.18k  231.36k    67.75%
  259304381 requests in 2.00m, 34.29GB read
Requests/sec: 2159239.24
Transfer/sec:    292.41MB
```

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (cf171ff) master (a48e5c7)

**Comment by trustin** [comment]
:+1:
